servant-server/src/Servant/Server/Internal/ErrorFormatter.hs

```diff
@@ -0,0 +1,64 @@
+{-# LANGUAGE DataKinds #-}
+
+-- | TODO: more documentation for this module
+module Servant.Server.Internal.ErrorFormatter
+  where
+
+import           Data.String.Conversions
+                 (cs)
+import           Data.Typeable
+
+import           Servant.API.ContentTypes
+import           Servant.Server.Internal.Context
+import           Servant.Server.Internal.ServerError
+
+-- | 'Context' that contains default formatters for all error types.
+--
+-- Default formatters will just return HTTP 400 status code with error
+-- message as response body.
+type DefaultErrorFormatters = '[BodyParseErrorFormatter, URLParseErrorFormatter, HeaderParseErrorFormatter]
+
+defaultErrorFormatters :: Context DefaultErrorFormatters
+defaultErrorFormatters =
+  defaulyBodyParseErrorFormatter
+  :. defaultURLParseErrorFormatter
+  :. defaultHeaderParseErrorFormatter
+  :. EmptyContext
+
+-- | A custom formatter for errors produced by parsing combinators like
+-- 'Servant.API.ReqBody' or 'Servant.API.Capture'.
+--
+-- A 'TypeRep' argument described the concrete combinator that raised
+-- the error, allowing formatter to customize the message for different
+-- combinators.
+type ErrorFormatter = TypeRep -> AcceptHeader -> String -> ServerError
+
+-- | Formatter for errors that occur while parsing request body.
+newtype BodyParseErrorFormatter = BodyParseErrorFormatter
+  { getBodyParseErrorFormatter :: ErrorFormatter
+  }
+
+defaulyBodyParseErrorFormatter :: BodyParseErrorFormatter
+defaulyBodyParseErrorFormatter = BodyParseErrorFormatter defaultErrorFormatter
+
+-- | Formatter for errors that occur while parsing URL parts, like 'Servant.API.Capture' or
+-- 'Servant.API.QueryParam'.
+newtype URLParseErrorFormatter = URLParseErrorFormatter
+  { getUrlParseErrorFormatter :: ErrorFormatter
+  }
+
+defaultURLParseErrorFormatter :: URLParseErrorFormatter
+defaultURLParseErrorFormatter = URLParseErrorFormatter defaultErrorFormatter
+
+-- | Formatter for errors that occur while parsing HTTP headers.
+newtype HeaderParseErrorFormatter = HeaderParseErrorFormatter
+  { getHeaderParseErrorFormatter :: ErrorFormatter
+  }
+
+defaultHeaderParseErrorFormatter :: HeaderParseErrorFormatter
+defaultHeaderParseErrorFormatter = HeaderParseErrorFormatter defaultErrorFormatter
+
+-- Internal
+
+defaultErrorFormatter :: ErrorFormatter
+defaultErrorFormatter _ _ e = err400 { errBody = cs e }
```

Comment:
I think it's a bit misleading to say this is the defaultErrorFormatter. This is more like `default400ErrorFormatter`. 400 might be thrown by more combinators than any other status code, but it's still not really a default.